File: successfulPairs.py
import math
class Solution:
    def binarySearch(self, target, potions):
        l,r=0,len(potions)
        while l<r:
            mid=l+(r-l)//2
            if target>potions[mid]:
                l=mid+1
            else:
                r=mid 
        return l
        
    def successfulPairs(self, spells, potions, success): # TC O(m+n(log m)) // SC O(m)
        n,m=len(spells), len(potions)
        potions.sort()
        ans=[0]*n
        for i in range(n):
            target=math.ceil(success/spells[i])
            idx=self.binarySearch(target, potions)
            #idx=bisect.bisect_left(potions, target)
            p=m-idx if m-idx>0 else 0
            ans[i]=p
        
        return ans

s=Solution()
print(s.successfulPairs([5,1,3],[1,2,3,4,5],7))

File: successfulPairs.java
import java.util.*;
// Approach 1- Sorting + Binary Search
class Solution { // TC O(m+n(log m)) SC O(log m)
    public int binarySearch(int target, int[] potions){
        int l=0, r=potions.length;
        while(l<r){
            int mid=l+(r-l)/2;
            if (target>potions[mid]){
                l=mid+1;
            } else {
                r=mid;
            }
        }
        return l;
    }
    
    public int[] successfulPairs(int[] spells, int[] potions, long success) {
        Arrays.sort(potions);
        int n=spells.length, m=potions.length;
        int[] answer=new int[n];
        for(int i=0; i<n; i++){
            int target=(int) Math.ceil((1.0*success)/spells[i]);
            int p=binarySearch(target, potions);
            int pairs= (m-p>0) ? (m-p): 0;
            answer[i]=pairs;
        }
        return answer;
    }
}

class successfulPairs{
    public static void main(String[] args) {
        int[] spells={5,1,3};
        int[] potions={1,2,3,4,5};
        long success=7;
        Solution s= new Solution();
        System.out.println(Arrays.toString(s.successfulPairs(spells, potions, success)));
    }
}